miniapps/mtop/mtop_integrators.hpp

```diff
@@ -0,0 +1,204 @@
+// Copyright (c) 2010-2020, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MTOPINTEGRATORS_HPP
+#define MTOPINTEGRATORS_HPP
+
+#include "mfem.hpp"
+#include <map>
+
+namespace mfem {
+
+
+class BaseQFunction
+{
+public:
+    virtual ~BaseQFunction(){}
+
+    virtual std::string GetType()=0;
+
+    //return the energy at a integration point
+    virtual
+    double QEnergy(ElementTransformation &T, const IntegrationPoint &ip,
+                   mfem::Vector &dd, mfem::Vector &uu)
+    {
+        return 0.0;
+    }
+
+    //return the residual at a integration point
+    virtual
+    void QResidual(ElementTransformation &T, const IntegrationPoint &ip,
+            mfem::Vector &dd, mfem::Vector &uu, mfem::Vector &rr)=0;
+
+    //return the gradient of the redidual at a integration point
+    virtual
+    void QGradResidual(ElementTransformation &T, const IntegrationPoint &ip,
+            mfem::Vector &dd, mfem::Vector &uu, mfem::DenseMatrix &hh)=0;
+
+    //return the gradient of the residual with respect to the design parameters, multiplied by the adjoint
+    virtual
+    void AQResidual(ElementTransformation &T, const IntegrationPoint &ip,
+                      mfem::Vector &dd, mfem::Vector &uu,
+                      mfem::Vector &aa, mfem::Vector &rr)=0;
+
+};
+
+
+class QLinearDiffusion:public BaseQFunction
```

Comment:
This class needs comments

Reply:
I have improved the comments - Additional comments will be added with the topopt miniapp